src/libraries/System.Private.CoreLib/src/System/IO/FileStatus.Unix.cs

```diff
@@ -402,23 +402,17 @@ internal void EnsureCachesInitialized(ReadOnlySpan<char> path, bool continueOnEr
         // Throws if any of the caches has an error number saved in it
         private void ThrowOnCacheInitializationError(ReadOnlySpan<char> path)
         {
-            int errno = 0;
-
             // Lstat should always be initialized by Refresh
             if (_initializedFileCache != 0)
             {
-                errno = _initializedFileCache;
+                InvalidateCaches();
+                throw Interop.GetExceptionForIoErrno(new Interop.ErrorInfo(_initializedFileCache), new string(path));
             }
             // Stat is optionally initialized when Refresh detects object is a symbolic link
             else if (_initializedSymlinkCache != 0 && _initializedSymlinkCache != -1)
-            {
-                errno = _initializedSymlinkCache;
-            }
-
-            if (errno != 0)
             {
                 InvalidateCaches();
-                throw Interop.GetExceptionForIoErrno(new Interop.ErrorInfo(errno), new string(path));
+                throw Interop.GetExceptionForIoErrno(new Interop.ErrorInfo(_initializedSymlinkCache), new string(path));
```

Comment:
because you took out the `if (errno != 0)` this code would now throw if `_initializedSymlinkCache` is 0 for some reason, whereas previously it would not.

Reply:
Line 412 should take care of that.
I think this would throw only for `_initializedSymlinkCache > 0`.

Reply:
Ah, I should have said: previously, if `_initializedSymlinkCache`==-1 and `_initializedFileCache` ==0, this method would not throw, and now it wil..

Reply:
no that's not right either. I'm tired. somehow you're passing -1 to ErrorInfo.

Reply:
Ah, it is caused by InvalidateCaches(), it is changing the value of _initialized* before passing it to ErrorInfo, I just sent the commit to avoid that.